pkg/asset/installconfig/aws/validation.go

```diff
@@ -0,0 +1,153 @@
+		}
+		if len(config.Compute) > 0 {
+			fldPath := field.NewPath("compute")
+			for idx, compute := range config.Compute {
+				fldPath := fldPath.Index(idx)
+				if compute.Platform.AWS != nil {
+					allErrs = append(allErrs, validateMachinePool(ctx, meta, fldPath.Child("platform", "aws"), config.Platform.AWS, compute.Platform.AWS)...)
+				}
+			}
+		}
+	}
+	return allErrs.ToAggregate()
+}
+
+func validatePlatform(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, networking *types.Networking, publish types.PublishingStrategy) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(platform.Subnets) > 0 {
+		allErrs = append(allErrs, validateSubnets(ctx, meta, fldPath.Child("subnets"), platform.Subnets, networking, publish)...)
+	}
+	return allErrs
+}
+
+func validateSubnets(ctx context.Context, meta *Metadata, fldPath *field.Path, subnets []string, networking *types.Networking, publish types.PublishingStrategy) field.ErrorList {
+	allErrs := field.ErrorList{}
+	privateSubnets, err := meta.PrivateSubnets(ctx)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, subnets, err.Error()))
+	}
+	privateSubnetsIdx := map[string]int{}
+	for idx, id := range subnets {
+		if _, ok := privateSubnets[id]; ok {
+			privateSubnetsIdx[id] = idx
+		}
+	}
+	if len(privateSubnets) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, "No private subnets found"))
+	}
+
+	publicSubnets, err := meta.PublicSubnets(ctx)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, subnets, err.Error()))
+	}
+	publicSubnetsIdx := map[string]int{}
+	for idx, id := range subnets {
+		if _, ok := publicSubnets[id]; ok {
+			publicSubnetsIdx[id] = idx
+		}
+	}
+
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, privateSubnets, privateSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, publicSubnets, publicSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, privateSubnets, privateSubnetsIdx, "private")...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, publicSubnets, publicSubnetsIdx, "public")...)
+
+	privateZones := sets.NewString()
+	publicZones := sets.NewString()
+	for _, subnet := range privateSubnets {
+		privateZones.Insert(subnet.Zone)
+	}
+	for _, subnet := range publicSubnets {
+		publicZones.Insert(subnet.Zone)
+	}
+	if publish != types.InternalPublishingStrategy && !publicZones.IsSuperset(privateZones) {
+		errMsg := fmt.Sprintf("No public subnet provided for zones %s", privateZones.Difference(publicZones).List())
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, errMsg))
+	}
+
+	return allErrs
+}
+
+func validateMachinePool(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, pool *awstypes.MachinePool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(pool.Zones) > 0 && len(platform.Subnets) > 0 {
```

Comment:
When `Subnets` is empty, do we want to take advantage of `meta` here and bring over [or current stopgap for `Zones` validation][1]?  It would be nice to compare vs. the output of `meta.AvailabilityZones()`.  Something like:

```go
if len(pool.Zones) {
  zones := sets.NewString(pool.Zones...)
  availableZones := sets.NewString()
  if len(platform.Subnets) > 0 {
    //pull from meta.PrivateSubnets
  } else {
    //pull from meta.AvailabilityZones
  }
  if diff := zones.Difference(availableZones); diff.Len() > 0 {
    //complain
  }
}
```
[1]: https://github.com/openshift/installer/blob/9b80f0c89ed47bceb8647571b02926e51144ec86/pkg/types/aws/validation/machinepool.go#L15-L19

Reply:
+1

Reply:
@wking fixed in https://github.com/openshift/installer/compare/a96182f3f4d7c7b69bd674bd918dd72a532f6ce4..9ce0bb39f5cbc2d581705794c8e16933cae350d4